### PR TITLE
[Feature] CrossGroupCritic — centralised critic spanning multiple age…

### LINKED ISSUE #3772: [Feature Request] CrossGroupCritic TensorDict Module for groups with different obs/action specs
##Solution + Motivation 
CrossGroupCritic — a single new TensorDictModule that acts as a drop-in critic. It does three things:

Per-group encoder (nn.Linear per group) — projects each group's observations to a shared embedding dim d. This is what handles heterogeneous obs shapes.
Shared MLP trunk — concatenates all groups' embeddings along the agent dim [B, n_a+n_b, d] and processes the joint representation.
Per-group output heads — splits the trunk output back by group size and writes a value key for each group.

##Why? 

MAPPOLoss doesn't change at all. You just pass CrossGroupCritic as the critic_network argument and call set_keys with the right group value key. Two MAPPOLoss instances (one per group) share the same critic object — gradients from both flow into it before each optimiser step.

Ad-hoc teamwork: add a detach_groups=["fixed_policy_group"] flag — the fixed group's encoder output is detached before the trunk, so the critic still sees their observations for a better baseline, but no gradient flows back into anything the fixed policy touches.

Scope: 1 new file (cross_group_critic.py), 2 export additions, 1 test file. Zero changes to existing MAPPO/IPPO code.

Fixes #3748

### PR DESCRIPTION
**Pull request by theap06**
Fixes #3772

Follow-up to the MAPPO/IPPO PR. Adds CrossGroupCritic, a TensorDictModule that centralises a value function across multiple heterogeneous agent groups — a use-case raised in review that single-group MultiAgentMLP cannot handle.

##Architecture:

Per-group encoders (Linear(obs_dim_g → d_model)) handle heterogeneous observation sizes without padding or alignment
Shared MLP trunk processes the joint team representation
Per-group value heads write estimates back to the tensordict

##Use cases:

Heterogeneous teams — agents in different groups have different obs/action specs
Ad-hoc teamwork — pass detach_groups=["fixed_group"] so a fixed-policy group informs the baseline without receiving gradients

##Changes:

torchrl/modules/models/cross_group_critic.py — new: GroupSpec dataclass + CrossGroupCritic
torchrl/modules/models/__init__.py, torchrl/modules/__init__.py — exports
test/test_modules.py — 13 tests: shapes, heterogeneous obs dims, shared/separate heads, gradient blocking via detach_groups, nested keys

### CODE REVIEW

File: test/modules/test_multiagent_models.py [hunk 1]
```diff
@@ -612,6 +612,189 @@ def test_mixer_malformed_input(
...
+
+        data = {}
+        for name, spec in group_map.items():
+            data[name] = {
+                "observation": torch.randn(*batch, spec.n_agents, spec.obs_dim)
+            }
+        return TensorDict(data, batch_size=list(batch))
+
+    def _make_group_map(self):
+        from torchrl.modules.models.cross_group_critic import GroupSpec
```
Comment (vmoens) [inline test/modules/test_multiagent_models.py]: Avoid local imports

File: test/modules/test_multiagent_models.py [hunk 2]
```diff
@@ -612,6 +612,189 @@ def test_mixer_malformed_input(
...
+        for name, spec in group_map.items():
+            val = td[name, "state_value"]
+            assert val.shape == (
+                *batch,
+                spec.n_agents,
+                1,
+            ), f"group '{name}': expected {(*batch, spec.n_agents, 1)}, got {val.shape}"
+
+    def test_heterogeneous_obs_dims(self):
+        from torchrl.modules.models.cross_group_critic import (
```
Comment (vmoens) [inline test/modules/test_multiagent_models.py]: Ditto

File: test/modules/test_multiagent_models.py [hunk 3]
```diff
@@ -612,6 +612,189 @@ def test_mixer_malformed_input(
...
+            },
+            batch_size=[4],
+        )
+        td = critic(td)
+        assert td["a", "val"].shape == (4, 2, 1)
+        assert td["b", "val"].shape == (4, 3, 1)
+        assert td["c", "val"].shape == (4, 1, 1)
+
+    def test_share_params_same_head(self):
+        from torchrl.modules.models.cross_group_critic import CrossGroupCritic
```
Comment (vmoens) [inline test/modules/test_multiagent_models.py]: Ditto

**Review by vmoens — APPROVED** [review]
LGTM thanks

**Comment by vmoens** [comment]
@theap06 just need to fix the local imports otherwise good to merge

**Comment by itwasabhi** [comment]
This looks excellent, thanks @theap06 -- will plan to use this soon!

File: torchrl/modules/__init__.py [hunk 4]
```diff
@@ -141,6 +144,7 @@
     "DuelingCnnDQNet",
     "EGreedyModule",
     "GPWorldModel",
+    "GroupSpec",
```
Comment (itwasabhi) [inline torchrl/modules/__init__.py]: Consider renaming to `CrossCriticGroupSpec`, as `GroupSpec` reads more general than it really is

**Comment by theap06** [comment]
> @theap06 just need to fix the local imports otherwise good to merge

Done

**Comment by theap06** [comment]
@vmoens flaky test. LGTM!

**Review by vmoens — APPROVED** [review]
LGTM
